controllers/chaosimpl/physicalmachinechaos/impl.go

```diff
@@ -0,0 +1,172 @@
+)
+
+type Impl struct {
+	client.Client
+	Log logr.Logger
+}
+
+func (impl *Impl) Apply(ctx context.Context, index int, records []*v1alpha1.Record, obj v1alpha1.InnerObject) (v1alpha1.Phase, error) {
+	impl.Log.Info("apply physical machine chaos")
+
+	physicalMachinechaos := obj.(*v1alpha1.PhysicalMachineChaos)
+	addresses := records[index].Id
+	addressArray := strings.Split(addresses, ",")
+
+	// for example, physicalMachinechaos.Spec.Action is 'network-delay', action is 'network', subAction is 'delay'
+	// notice: 'process' action has no subAction, set subAction to ""
+	actions := strings.SplitN(string(physicalMachinechaos.Spec.Action), "-", 2)
+	if len(actions) == 1 {
+		actions = append(actions, "")
+	} else if len(actions) != 2 {
+		err := errors.New("action invalid")
+		return v1alpha1.NotInjected, err
+	}
+	action, subAction := actions[0], actions[1]
+	physicalMachinechaos.Spec.ExpInfo.Action = subAction
+
+	/*
+		transform ExpInfo in PhysicalMachineChaos to json data required by chaosd
+		for example:
+		    ExpInfo: &ExpInfo {
+			    UID: "123",
+				Action: "cpu",
+				StressCPU: &StressCPU {
+					Load: 1,
+					Workers: 1,
+				}
+			}
+
+			transform to json data: "{\"uid\":\"123\",\"action\":\"cpu\",\"load\":1,\"workers\":1}
+	*/
+	var expInfoMap map[string]interface{}
+	expInfoBytes, _ := json.Marshal(physicalMachinechaos.Spec.ExpInfo)
+	err := json.Unmarshal(expInfoBytes, &expInfoMap)
+	if err != nil {
+		impl.Log.Error(err, "fail to unmarshal experiment info")
+		return v1alpha1.NotInjected, err
+	}
+	configKV, ok := expInfoMap[string(physicalMachinechaos.Spec.Action)].(map[string]interface{})
+	if !ok {
+		err = errors.New("transform action config to map failed")
+		impl.Log.Error(err, "")
+		return v1alpha1.NotInjected, err
+	}
+	for k, v := range configKV {
+		expInfoMap[k] = v
+	}
+	delete(expInfoMap, string(physicalMachinechaos.Spec.Action))
+
+	expInfoBytes, err = json.Marshal(expInfoMap)
+	if err != nil {
+		impl.Log.Error(err, "fail to marshal experiment info")
+		return v1alpha1.NotInjected, err
+	}
+
+	for _, address := range addressArray {
+		url := fmt.Sprintf("%s/api/attack/%s", address, action)
+		impl.Log.Info("HTTP request", "address", address, "data", string(expInfoBytes))
+
+		if err := impl.doHttpRequest("POST", url, bytes.NewBuffer(expInfoBytes)); err != nil {
+			return v1alpha1.Injected, err
+		}
+	}
+
+	return v1alpha1.Injected, nil
+}
+
+func (impl *Impl) Recover(ctx context.Context, index int, records []*v1alpha1.Record, obj v1alpha1.InnerObject) (v1alpha1.Phase, error) {
+	impl.Log.Info("recover physical machine chaos")
+
+	physicalMachinechaos := obj.(*v1alpha1.PhysicalMachineChaos)
+	addresses := records[index].Id
+
+	addressArray := strings.Split(addresses, ",")
+	for _, address := range addressArray {
+		url := fmt.Sprintf("%s/api/attack/%s", address, physicalMachinechaos.Spec.ExpInfo.UID)
+		if err := impl.doHttpRequest("DELETE", url, nil); err != nil {
+			return v1alpha1.Injected, err
+		}
+	}
```

Comment:
The logic of recover is not idempotent, because it seems chaosd restful API `DELETE /api/attack/<uid>` does not resolve "not exists" very well, it will return 500 but not 404 when the experiment has already been deleted. 

For example:
- there are three addresses (addrA, addrB, and addrC) in this physical machine chaos
- `DELETE /api/attack/<uid>` succeed on addrA and addrB, but failed on addrC (connection refused, or return 50x);
- line 119, return Injected with error, that would trigger the retry 
- in the next reconciliations, DELETE request on addrA and addrB would keep becoming failed, so this experiment would never be recovered.

I think the best way is returning 404 in chaosd, if the experiment has been already deleted, and resolve the 404 out of `doHttpRequest` (by return certain `NotExist` error, or more return values). But this way relates many changes.

Or checking the experiment exists or not before deleting it.

Reply:
Good idea, I update in https://github.com/WangXiangUSTC/chaos-mesh/commit/21f142103717e7e2839aeb33dc5e980285b2468b, and create an issue in chaosd https://github.com/chaos-mesh/chaosd/issues/93